File: src/shared/golang/backend/controls/controls.go
package controls

import (
	"github.com/jmoiron/sqlx"
)

type ControlFrequencyType int32
type ControlType int32

const (
	CFTAdHoc    ControlFrequencyType = 1
	CFTInterval                      = 2
	CFTOther                         = 3
)

const (
	CTAccess           ControlType = 1
	CTAuthorization                = 2
	CTConfiguration                = 3
	CTGitc                         = 4
	CTInterface                    = 5
	CTManagementReview             = 6
	CTReconciliation               = 7
	CTReport                       = 8
)

type Control struct {
	Id                      int64                `db:"id"`
	EngagementId            int64                `db:"engagement_id"`
	Name                    string               `db:"name"`
	HumanId                 string               `db:"human_id"`
	Likelihood              int16                `db:"likelihood"`
	LikelihoodJustification string               `db:"likelihood_justification"`
	ControlType             ControlType          `db:"control_type"`
	OwnerId                 *int64               `db:"owner_id"`
	IsManual                bool                 `db:"is_manual"`
	FreqType                ControlFrequencyType `db:"freq_type"`
	FreqOther               string               `db:"freq_other"`
	FreqInterval            string               `db:"freq_interval"`
	Description             string               `db:"description"`
}

type ControlManager struct {
	db *sqlx.DB
}

func CreateControlManager(db *sqlx.DB) *ControlManager {
	return &ControlManager{
		db: db,
	}
}

File: src/webapp/client/src/ts/types/controls.ts
export enum ControlFrequencyType {
	CFTAdHoc     = 1,
	CFTInterval  = 2,
	CFTOther     = 3,
}

export enum ControlType {
	CTAccess           = 1,
	CTAuthorization    = 2,
	CTConfiguration    = 3,
	CTGitc             = 4,
	CTInterface        = 5,
	CTManagementReview = 6,
	CTReconciliation   = 7,
	CTReport           = 8,
}

export interface RawControl {
    Id                     : number
    EngagementId           : number
	Name                   : string
	HumanId                : string
	Likelihood             : number
	LikelihoodJustification: string
	ControlType            : ControlType
	OwnerId                : number | null
	IsManual               : boolean
	FreqType               : ControlFrequencyType
	FreqOther              : string
	FreqInterval           : string
	Description            : string
}

export function createEmptyControl() : RawControl {
    return {
        Id: -1,
        EngagementId: -1,
        Name: '',
        HumanId: '',
        Likelihood: 0,
        LikelihoodJustification: '',
        ControlType: ControlType.CTAccess,
        OwnerId: null,
        IsManual: true,
        FreqType: ControlFrequencyType.CFTAdHoc,
        FreqOther: '',
        FreqInterval: '',
        Description: '',
    }
}
